src/SignalR/clients/java/signalr/src/main/java/com/microsoft/signalr/MessagePackHubProtocol.java

```diff
@@ -0,0 +1,617 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+package com.microsoft.signalr;
+
+import java.io.IOException;
+import java.lang.reflect.Type;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import org.msgpack.core.MessageBufferPacker;
+import org.msgpack.core.MessageFormat;
+import org.msgpack.core.MessagePack;
+import org.msgpack.core.MessagePackException;
+import org.msgpack.core.MessagePacker;
+import org.msgpack.core.MessageUnpacker;
+import org.msgpack.jackson.dataformat.MessagePackFactory;
+import org.msgpack.value.ValueType;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.type.TypeFactory;
+
+public class MessagePackHubProtocol implements HubProtocol {
+    
+    private static final int ERROR_RESULT = 1;
+    private static final int VOID_RESULT = 2;
+    private static final int NON_VOID_RESULT = 3;
+    
+    private ObjectMapper objectMapper = new ObjectMapper(new MessagePackFactory());
+    private TypeFactory typeFactory = objectMapper.getTypeFactory();
+
+    @Override
+    public String getName() {
+        return "messagepack";
+    }
+
+    @Override
+    public int getVersion() {
+        return 1;
+    }
+
+    @Override
+    public TransferFormat getTransferFormat() {
+        return TransferFormat.BINARY;
+    }
+
+    @Override
+    public List<HubMessage> parseMessages(ByteBuffer payload, InvocationBinder binder) {
+        if (payload.remaining() == 0) {
+            return null;
+        }
+
+        List<HubMessage> hubMessages = new ArrayList<>();
+        
+        try {
+            while (payload.hasRemaining()) {
+                int length = Utils.readLengthHeader(payload);
+                // Throw if remaining buffer is shorter than length header
+                if (payload.remaining() < length) {
+                    throw new RuntimeException(String.format("MessagePack message was length %d but claimed to be length %d.", payload.remaining(), length));
+                }
+                // Instantiate MessageUnpacker
+                MessageUnpacker unpacker = MessagePack.newDefaultUnpacker(payload);
```

Comment:
Does this implement closeable? If so we should use try { } with it so it closes even with exceptions

Reply:
It implements autocloseable, I'll convert to a try-with-resources